Added system tray mode to wireguard service executable

spylogsster opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/31142

- Added interactive mode with system tray menu to manage vpn connection. Synchronization of menu items state with the service and the browser will be added in next PR, here is UI side only.
- Moved status icon related code into status_icon  folder
- Renamed interactive to status_tray
- Renamed InteractiveMain to StatusTrayRunner
- Moved service_constants.h/cc to `//brave/components/brave_vpn/common/wireguard`

<img width="209" alt="image" src="https://github.com/brave/brave-core/assets/2965009/e4bdb951-fc4c-4a5d-9863-b51db6e04af9">

<img width="209" alt="image" src="https://github.com/brave/brave-core/assets/2965009/fb4910ed-8124-44b4-b6d2-cc52746d9bca">


<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-arm64, CI/skip-linux-x64, CI/skip-android, CI/skip-macos, CI/skip-ios, CI/skip-windows-arm64, CI/skip-windows-x64, CI/skip-windows-x86 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- no manual tests required in this PR, all tests will be in https://github.com/brave/brave-browser/issues/30708

bridiver commented on components/brave_vpn/browser/connection/common/win/utils.h:
```diff
@@ -14,6 +14,7 @@
 namespace brave_vpn {
```
There should not be a `common` directory under `browser`. `common` has a specific meaning in chromium which is files that are shared across `browser`, `renderer`, etc...  so it doesn't make sense to have `common` within `browser`

spylogsster replied:
same as in other cases

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/common/service_constants.h:
```diff
@@ -26,6 +26,10 @@ constexpr char kBraveVpnWireguardServiceUnnstallSwitchName[] = "uninstall";
 // Load wireguard binaries and connect to vpn using passed config.
 constexpr char kBraveVpnWireguardServiceConnectSwitchName[] = "connect";
```
same as `common/win`, this is an incorrect usage of `common` in chromium

spylogsster replied:
this is separate executable and common code inside it

bridiver replied:
Common to what? I think you are using "common" here with a different meaning and we should not be doing that because it's confusing in the context of chromium code. If it's common to the browser process code and exe then it should go in `brave_vpn/common/wireguard` or something

bridiver replied:
and for instance here https://github.com/brave/brave-core/pull/18960/files#diff-032ee8426b314ef9f20a5b383cd195294523101c75b533685d14d2665e8ce71fR1 having `common` makes no sense because that code is only used for status_tray

spylogsster replied:
sure, reworking of `components/brave_vpn/browser/connection/common`  and `components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/common`  is out of scope for this PR and affects many files across browser process, ikev2 helper process and wireguard service preocess, I will take it next as separate PR in follow up https://github.com/brave/brave-browser/issues/31217

This one was added here `components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/common` and I renamed it to utils.

bridiver replied:
Moving a file is not "out of scope". This is a very simple change to correct a problem with a file you made a change to in this PR.

spylogsster replied:
@bridiver done here https://github.com/brave/brave-core/pull/18960/commits/eab1cddf23d4ec17f4728999a810967efbba6809

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/interactive/common/scoped_hmenu.h:
```diff
@@ -0,0 +1,36 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_VPN_BROWSER_CONNECTION_WIREGUARD_WIN_BRAVE_VPN_WIREGUARD_SERVICE_INTERACTIVE_COMMON_SCOPED_HMENU_H_
```
see others re: common

bridiver replied:
this isn't common to anything anyway, it's only used in StatusTrayWin. I would just define it there.

spylogsster replied:
this is separate executable and common code inside it

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/interactive/interactive_utils.cc:
```diff
@@ -0,0 +1,65 @@
+}
+
+bool UseDarkTheme() {
+  base::win::RegKey key;
```
why is this not using the chromium code for dark mode?

spylogsster replied:
discussed in slack + this function does same actually

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/interactive/interactive_main.cc:
```diff
@@ -0,0 +1,141 @@
+constexpr char kBraveAccountURL[] = "http://account.brave.com/";
+constexpr char kAboutBraveVPNURL[] = "https://brave.com/firewall-vpn/";
+
+std::u16string GetStatusIconTooltip(bool connected, bool error) {
```
this stuff is all for the tray, right? Why are we not using the system/tray code in chromium?

spylogsster replied:
discussed in slack

bridiver commented:
It seems like we have a lot of custom code in here for handling a tray icon instead of using the chromium helpers
```
// Provides a cross-platform interface to the system's status tray, and exposes
// APIs to add/remove icons to the tray and attach context menus.
class StatusTray {
```

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/interactive/interactive_main.cc:
```diff
@@ -0,0 +1,141 @@
+
+InteractiveMain::~InteractiveMain() = default;
+
+void InteractiveMain::SetupStatusIcon() {
```
why is this called `*Main`? That convention is normally used for `main` or equivalent entry point for an executable?

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/interactive/common/scoped_hwnd.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "base/scoped_generic.h"
+
+namespace brave {
```
This namespace should match the component

spylogsster replied:
done

spylogsster commented:
Added the list of refactorings after our Slack discussion with @bridiver, here the list of changes:
- Moved status icon related code into status_icon  folder
- Renamed interactive to status_tray
- Renamed InteractiveMain to StatusTrayRunner
- Moved wireguard service code to service  folder 
- Renamed ServiceMain to WireguardServiceRunner
- Added some small local refactorings due to new changes and more comments .

spylogsster commented:
> It seems like we have a lot of custom code in here for handling a tray icon instead of using the chromium helpers
> 
> ```
> 
> // Provides a cross-platform interface to the system's status tray, and exposes
> 
> // APIs to add/remove icons to the tray and attach context menus.
> 
> class StatusTray {
> 
> ```

Regarding the reuse of chromium source code from chrome/browser/status_icons/ and  chrome/browser/ui/views/status_icons/ I tried to do this as the first step, but encountered a number of problems. Some of them managed to solve by adding "local hacks", but still ended up in a crash on ubrk_open, which in a reasonable time could not be solved.
Even if this problem is solved, the benefit of reusing that code is not obvious, it will only get rid of 3 classes, which in itself is not big, but will add dependencies in chrome, bloat the process initialization code and require periodic refactoring for each merge due to changes in chrome.
We discussed it with and agreed to keep current implementation

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/status_icon/native_popup_menu.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace brave_vpn {
+
+class NativePopupMenu {
```
Q) chromium has `NativeMenuWin`. Is it not re-usable here?

spylogsster replied:
yep, it doesn't provide access to HMENU so it is impossible to pass it to TrackPopupMenu

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/utils/utils.h:
```diff
@@ -0,0 +1,20 @@
+
+namespace brave_vpn {
+bool UseDarkTheme();
+gfx::ImageSkia GetIconFromResources(int icon_id, gfx::Size size);
```
```suggestion
gfx::ImageSkia GetIconFromResources(int icon_id, const gfx::Size& size);
```

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/utils/utils.cc:
```diff
@@ -0,0 +1,65 @@
+}
+}  // namespace
+
+gfx::ImageSkia GetIconFromResources(int icon_id, gfx::Size size) {
```
```suggestion
gfx::ImageSkia GetIconFromResources(int icon_id, const gfx::Size& size) {
```

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/status_tray_runner.cc:
```diff
@@ -0,0 +1,176 @@
+
+namespace {
+
+constexpr char kBraveAccountURL[] = "http://account.brave.com/";
```
```suggestion
constexpr char kBraveAccountURL[] = "https://account.brave.com/";
```

simonhong replied:
btw, can't we reuse these urls from components\brave_vpn\common\brave_vpn_constants.h?

spylogsster replied:
good idea, thanks, updated

simonhong reviewed (approved):
++ with trivial nits.

bridiver commented on components/brave_vpn/DEPS:
```diff
@@ -6,4 +6,5 @@ include_rules = [
   "+net",
   "+url",
   "+brave/components/api_request_helper",
+  "+chrome/install_static",
```
I totally missed this before, you can't access `chrome/*` from a component, that's a layering violation. Also DEPS should always be defined only for the directories they are actually needed in so this file shouldn't exist at all. base, net, url, etc... are always accessible for components so they don't need to be here at all. You should only add DEPS in the specific directory that needs those additional include rules

spylogsster replied:
moved as discussed in Slack and created follow up https://github.com/brave/brave-browser/issues/31306

spylogsster replied:
follow up for methods https://github.com/brave/brave-browser/issues/31307

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/BUILD.gn:
```diff
@@ -30,8 +30,10 @@ executable("brave_vpn_wireguard_service") {
     "//base:base_static",
+    "//brave/components/brave_vpn/common/wireguard",
     "//build/win:default_exe_manifest",
     "//chrome/install_static:install_static_util",
```
how did this ever get in here before you added https://github.com/brave/brave-core/pull/18960/files#diff-89ddcd9ed3c4baf29fffec235e5062d0c5c7c4694c00b99ed985e38e4f322d6dR9 ?
This is explicitly not allowed here

spylogsster replied:
discussed in Slack and created follow up https://github.com/brave/brave-browser/issues/31306

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+    "status_tray_runner.h",
+  ]
+
+  public_deps = [ "//skia" ]
```
why is this a public dep?

spylogsster replied:
because skia works only as public

spylogsster replied:
discussed in Slack and replaced //skia by //ui/base

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/status_icon/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+    "tray_menu_model.h",
+  ]
+
+  public_deps = [ "//skia" ]
```
why do you have all these public deps for `skia`? You don't even have any direct dependencies on skia so this is definitely not right

spylogsster replied:
ImageModel depends on skia and it doesnt work as private dep

spylogsster replied:
replaced //skia by //ui/base

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/utils/utils.cc:
```diff
@@ -0,0 +1,65 @@
+  return family->CreateExact(size).AsImageSkia();
+}
+
+bool UseDarkTheme() {
```
use `//ui/native_theme`

spylogsster replied:
done

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+    "//ui/gfx",
+  ]
+
+  libs = [ "Shell32.lib" ]
```
`Shell32.lib` -> `shell32.lib`

spylogsster replied:
done

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+    "status_tray_runner.h",
+  ]
+
+  public_deps = [ "//ui/base" ]
```
this shouldn't need to be a public dep here because it's not exposed in the header

bridiver replied:
actually it doesn't appear to be used at all in this target

spylogsster replied:
`TrayMenuModel used in header -> ui/base/simple_menu_model.h -> ...`

bridiver replied:
Never add a transitive dep. If you're seeing an issue always because of a problem with a direct dep. Fix the root of the issue (which might mean making a direct dep a public_dep)

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+    "//brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/common",
+    "//brave/components/brave_vpn/common",
+    "//brave/components/brave_vpn/common/wireguard/win",
+    "//ui/gfx",
```
`ui/gfx` is not used in this target as far as I can tell

bridiver replied:
also remove from status_tray/DEPS

spylogsster replied:
done

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+  deps = [
+    "resources",
+    "status_icon",
+    "utils",
```
This dep does not appear to be used in this target

spylogsster replied:
status_tray_runner.cc uses it

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+    "//brave/components/brave_vpn/common/wireguard/win",
+    "//ui/gfx",
+    "//ui/native_theme",
+    "//ui/native_theme:native_theme_browser",
```
This dep does not appear to be used here. Probably the best thing is to remove all the deps from these BUILD.gn files and then re-add the ones that gn complains about

spylogsster replied:
"//ui/native_theme" cannot be included without `native_theme_browser`, it throws errors, only both work

bridiver replied:
strange, but other uses have the same thing so I guess it's ok

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/status_icon/native_popup_menu.h:
```diff
@@ -0,0 +1,45 @@
+
+#include "base/memory/raw_ptr.h"
+#include "brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/utils/scoped_hmenu.h"
+#include "ui/base/models/menu_model.h"
```
this could be a forward declaration and that should eliminate your public_dep issue entirely

spylogsster replied:
added forward declaration

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/status_icon/BUILD.gn:
```diff
@@ -0,0 +1,27 @@
+
+  deps = [
+    "//base",
+    "//ui/base",
```
this should be a public_dep and then you can remove the public_dep above that is not a direct dependency

spylogsster replied:
done
